boutdata/data.py

```diff
@@ -1034,8 +1131,60 @@ def __init__(
                 self._datacachesize = 0
                 self._datacachemaxsize = self._caching * 1.0e9
 
+        if self._parallel is not False:
+            if self._parallel is True or self._parallel == 0:
+                self._parallel = determineNumberOfCPUs()
+            if not isinstance(self._parallel, int) or self._parallel <= 0:
+                raise ValueError(
+                    "Passed or found inconsistent value %i for number of processes",
+                    self._parallel,
+                )
+
+            # Open the 0'th file so we can read scalars without the worker processes
+            self._root_file = DataFile(self._file_list[0])
+
+            # Need to initialise all workers with a shared memory buffer to write to
+            dim_sizes = tuple(self.sizes[d] for d in ("t", "x", "y", "z"))
+            self._shared_buffer_raw = RawArray("d", int(numpy.prod(dim_sizes)))
+            self._shared_buffer = numpy.reshape(
+                numpy.frombuffer(self._shared_buffer_raw), dim_sizes
+            )
+
+            # Work out which files to assign to which workers
+            min_files_per_proc = int(self.npes) // self._parallel
+            extra_files = int(self.npes) % self._parallel
+            files_per_proc = [min_files_per_proc] * (self._parallel - extra_files) + [
+                min_files_per_proc + 1
+            ] * extra_files
+            # self._workers is a list of pairs of (worker, connection)
+            self._workers = []
+            filenum = 0
+            for i in range(self._parallel):
+                parent_connection, child_connection = Pipe()
+                proc_list = tuple(
+                    p for p in range(filenum, filenum + files_per_proc[i])
+                )
+                filenum = filenum + files_per_proc[i]
+                self._workers.append(
+                    (
+                        Process(
+                            target=self._worker_function,
+                            args=(child_connection, proc_list, self._shared_buffer_raw),
+                        ),
+                        parent_connection,
+                    )
+                )
+                self._workers[-1][0].start()
```

Comment:
The constructor is quite long.
Would this be a good time to split it to more (private) sub-routines in order to improve maintainability?
For example it looks to code under `if self._parallel is not False:` could be made into a sub-routine.